crates/net/types/src/trusted_peer.rs

```diff
@@ -0,0 +1,354 @@
+//! `NodeRecord` type that uses a domain instead of an IP.
+
+use std::{
+    fmt::{self, Write},
+    io::Error,
+    net::IpAddr,
+    num::ParseIntError,
+    str::FromStr,
+};
+
+use crate::{NodeRecord, PeerId};
+use secp256k1::{SecretKey, SECP256K1};
+use serde_with::{DeserializeFromStr, SerializeDisplay};
+use std::time::Duration;
+use tokio_retry::{strategy::FixedInterval, Retry};
+use url::Host;
+
+/// Represents the node record of a trusted peer. The only difference between this and a
+/// [`NodeRecord`] is that this does not contain the IP address of the peer, but rather a domain
+/// __or__ IP address.
+///
+/// This is useful when specifying nodes which are in internal infrastructure and may only be
+/// discoverable reliably using DNS.
+///
+/// This should NOT be used for any use case other than in trusted peer lists.
+#[derive(Clone, Debug, Eq, PartialEq, Hash, SerializeDisplay, DeserializeFromStr)]
+pub struct TrustedPeer {
+    /// The host of a node.
+    pub host: Host,
+    /// TCP port of the port that accepts connections.
+    pub tcp_port: u16,
+    /// UDP discovery port.
+    pub udp_port: u16,
+    /// Public key of the discovery service
+    pub id: PeerId,
+}
+
+/// Retry strategy for DNS lookups.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct RetryStrategy {
+    /// The amount of time between attempts.
+    pub interval: Duration,
+    /// The number of attempts to make before failing.
+    pub attempts: usize,
+}
+
+impl RetryStrategy {
+    /// Create a new retry strategy.
+    pub const fn new(interval: Duration, attempts: usize) -> Self {
+        Self { interval, attempts }
+    }
+}
+
+impl TrustedPeer {
+    /// Derive the [`NodeRecord`] from the secret key and addr
+    pub fn from_secret_key(host: Host, port: u16, sk: &SecretKey) -> Self {
+        let pk = secp256k1::PublicKey::from_secret_key(SECP256K1, sk);
+        let id = PeerId::from_slice(&pk.serialize_uncompressed()[1..]);
+        Self::new(host, port, id)
+    }
+
+    /// Creates a new record from a socket addr and peer id.
+    pub const fn new(host: Host, port: u16, id: PeerId) -> Self {
+        Self { host, tcp_port: port, udp_port: port, id }
+    }
+
+    /// Resolves the host in a [`TrustedPeer`] to an IP address, returning a [`NodeRecord`].
+    ///
+    /// This takes as input a [`RetryStrategy`] which specifies how to handle DNS lookup failures.
+    ///
+    /// Using a retry strategy of `None` will not retry the lookup if a lookup is made and it fails.
+    pub async fn resolve(
+        &self,
+        retry_strategy: Option<RetryStrategy>,
+    ) -> Result<NodeRecord, Error> {
```

Comment:
I don't like intregrating a retrystrategy in there like this.

imo it should be the caller's responsibility, to retry the resolve fn.

I'd rather have this as a separate function, or leave it up to the caller entirely.

we also already use the backon crate for retries. can we use that instead, the tokio-retry crate is very smol, but still. Also open to remove backon in favor of tokio-retry

Reply:
Have created #8613 to address this 🙏

Reply:
I can't see `load_toml_config()` being tested. Now that the retry logic is moved there, I feel obligated to add tests for that fn.

Is that functionality perhaps tested by some e2e suite in CI? Or would we want to add a unit test for it perhaps?